### ISSUE TITLE
Cast Android TV Receiver - Correct implementation with Media3

### ISSUE THREAD

**Issue by TomasValenta** [androidx/media#848]
What is the current best approach to implement the Cast Android TV Receiver SDK with Media3?
- I did not find any useful information on this to date.
- The Media3 Demo project directory contains only Cast Sender example.
- ExoPlayer project related references [Cast docs](https://developers.google.com/cast/docs/android_tv_receiver/core_features#using_exoplayer_with_cast_connect), [CastAndroidTvReceiver example](https://github.com/googlecast/CastAndroidTvReceiver/), [Marc Medium post](https://medium.com/google-exoplayer/the-mediasession-extension-for-exoplayer-82b9619deb2d) are all outdated, deprecated and they no longer exist in Media3 (ie `MediaSessionConnector`).
- Based on findings from `CastPlayer` source code, it always uses `remoteMediaClient.queueLoad` method, which requires a specific approach [on Receiver side](https://developers.google.com/cast/docs/android_tv_receiver/queueing#handle_queue_loading_and_setting_the_queue_information).

I have a very basic implementation done at the moment, but I don't know exactly when and what to broadcast with the MediaStatus differently from the official [example mentioned above](https://github.com/googlecast/CastAndroidTvReceiver/blob/master/app-kotlin/src/main/kotlin/com/google/sample/cast/atvreceiver/ui/PlaybackVideoFragment.kt).

For example `(int) currentTimeline.getPeriod(mediaItemIndex, period).uid` in CastPlayer `seekTo` method returns (after disconnection and after new connection) wrong value (exceeding the queue size). But this works fine with the Default Web Receiver App, so from this I conclude that I am missing something to properly integrate on the Receiver side.

Is there an official demo planned, like is there the one for Cast Sender?

**Comment by marcbaechinger** [comment]
Like you state, the `CastPlayer` uses the `RemoteMediaClient` to prepare the player with a queue and send playback commands. These messages are then sent by the Cast SDK to the receiver SDK on the cast device.

In the manifest of the cast demo app is a [cast options provider registered](https://github.com/androidx/media/blob/release/demos/cast/src/main/AndroidManifest.xml#L34). It uses `DefaultCastoptionsProvider` provided by the library that [declares the app id to be started](https://github.com/androidx/media/blob/release/libraries/cast/src/main/java/androidx/media3/cast/DefaultCastOptionsProvider.java#L37).

You can do your own receiver and change that ID to your app ID.

I don't know much about the receiver side API to be honest. The web receiver SDK [gives you a &lt;cast-media-player&gt; element](https://developers.google.com/cast/docs/web_receiver/basic#create_a_basic_web_receiver_app) and API components like [a context, a `PlayerManager` and friends](https://developers.google.com/cast/docs/web_receiver/core_features#configure_the_player). Your app basically displays that HTML-element and the playback commands sent by `CastPlayer` through `RemoteMediaClient` are received and applied by the SDK on the receiver end.  The rest around that is your HTML/JS app.

From what I know, `CastPlayer` works with the default receiver, a styled receiver or such a custom receiver that uses the Cast web SDK. 

For more information on how to implement such a receiver we can't help you here in much more detail I'm afraid. The links above should be a starting point.

**Comment by TomasValenta** [comment]
First thank you for your attention Marc,

I know you get a lot of questions every day and I'm sorry, I probably didn't describe my question well enough. I'm working on integration Cast [Android TV Receiver App](https://developers.google.com/cast/docs/android_tv_receiver) (not the Web Receiver App):

![image](https://github.com/androidx/media/assets/4901251/795ace22-8fe7-4eba-bf30-e16251cff846)
 
And as I mentioned above, I have already completed the basic integration and it partially works, but I have already spent a lot of time on some edge cases and I still cannot solve some of them.

That's why I asked if you are planning any official demo / integration support, like you have for the Cast Sender part (the CastPlayer)? 

Btw: The [link to your medium post](https://medium.com/google-exoplayer/the-mediasession-extension-for-exoplayer-82b9619deb2d) with the old `MediaSessionConnector` I mentioned because it is listed as a reference link in the [code lab for ATV app integration](https://developers.google.com/cast/codelabs/cast-connect-atv#3) at the very bottom of the page 4 🙂

![image](https://github.com/androidx/media/assets/4901251/c8557a1d-e891-4bd6-a747-84a2509042ce)

**Comment by marcbaechinger** [comment]
Yeah, sorry I'm not familiar with this. That's an app published by Cast folks. We weren't involved. 

I can just point to this page that references a code lab and a sample app, but you are probably better informed about these resources than I am. 

We don't have plans to provide a Android TV receiver app any time soon.
Sorry for not having a better answer.

@pflammertsma or @AquilesCanta  may have some thoughts.

**Comment by TRtomasz** [comment]
@marcbaechinger  @AquilesCanta @pflammertsma
hello guys, any update regarding this?
@TomasValenta  were you able to figure it out on your own?

**Comment by TomasValenta** [comment]
Sorry for the late reply! We’ve decided to postpone this feature for now and continue using the `Default Web Receiver App`.

**Comment by marcbaechinger** [comment]
Thanks! I'm closing mostly because I can't really help. The cast team would need to help here I think.

**Comment by AquilesCanta** [comment]
Apologies, I hadn't seen this before.

Media3, through the cast extension, only supports sender-side integration as of today. Which means this is not really the right place to ask about CastConnect / AndroidTV native receiver. I think the right place to file this would be https://issuetracker.google.com/issues/new?component=190205&template=814901.

I have some plans to work on CastConnect integration functionality in the context of the Cast extension in media3, but I cannot give you any specific timelines just yet, as we are still working on improving the sender side functinoality. Hopefully you've had a chance to try the latest Cast extension changes?

So for the time being, I think the link that Marc provided above is the right place https://developers.google.com/cast/docs/android_tv_receiver/core_features. Otherwise, hopefully, we'll be able to give you a better answer in 2026, so that you can launch your native Android app when casting to an Android TV.
